### DISCUSSION TITLE
IRAF on Mac M2?

### DISCUSSION THREAD

**Discussion by Mark-Dickinson** [iraf-community/iraf#257] [General]
Hello – kudos for this IRAF community effort!  Reading the [v2.17 release notes](https://github.com/iraf-community/iraf/releases/tag/v2.17) I read that "Mac on ARM64 (M1) is supported as well", and indeed a colleague reported that he had successfully installed IRAF on an M1 Mac Pro.  Now there are new M2 Mac Pros, and I was wondering if there's an effort to support IRAF on M2s as well, or if that has already been tried (successfully or not!).

Thank you - Mark

**Comment by olebole** [comment]
I made a short test on a new Macbook with macOS 13 Ventura and it seems to work well. The M2 is compatible to the M1.

**Reply by Mark-Dickinson** [reply]
Thank you!   So perhaps I'll get that new MacBook after all...

**Reply by pgkalas** [reply]
Mark, my IRAF 2.17 installation seems to work on a new M2 MacBook pro with Ventura.  Note, however, external packages such as [stsdas](https://github.com/spacetelescope) have not been converted to 64-bit.  You can find the python packages [here](https://github.com/spacetelescope).

**Reply by Mark-Dickinson** [reply]
Thank you, good to hear!  I was aware that STSDAS had not been suitably renovated – not too surprising, c'est la vie.  We're using STScI tools with JWST data, but sometimes dinosaurs like me still need that ol' IRAF magic...

(BTW is that you, Paul?  If so, hi! and if not, sorry, never mind!)

**Reply by pgkalas** [reply]
Hi Mark, yes it's me!  Glad to e-connect after so many years!

**Comment by ian-smail** [comment]
Hi: I've just tried to do an install on an M2 MBP running Monterey and have run into a problem:  

+ gcc generic.o lexyy.o yywrap.o /usr/local/iraf/iraf-2.17/unix/hlib/libboot.a /usr/local/iraf/iraf-2.17/lib/libsys.a /usr/local/iraf/iraf-2.17/lib/libvops.a /usr/local/iraf/iraf-2.17/unix/hlib/libos.a /usr/local/iraf/iraf-2.17/unix/bin.macos64/libf2c.a -lm -o generic.e
ld: file too small (length=8) file '/usr/local/iraf/iraf-2.17/lib/libsys.a' for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [sysgen] Error 1

any advice would be welcome!  build log is attached.  

Thanks - Ian

[build.log](https://github.com/iraf-community/iraf/files/10892201/build.log)

**Reply by olebole** [reply]
The reason is this:
```
xc: Command not found.
```
in the log file. xc is built before, and copied  `unix/bin.macos64` according to your log. If it is not there, then something with your directory structure is wrong. If it is there, make sure that in `$HOME/.iraf/bin/` contains a link `xc --> /usr/local/iraf/iraf-2.17/unix/bin.macos64/xc.e`. This link should have been created when you have run `./install` as the first step.

**Reply by ian-smail** [reply]
so repeating the build from scratch and just adopting the defaults for the installation - i end up with an executable version of xc (see attached screen grab) - but the make sysgen script still falls over (build.log.2).  

rerunning the build doesn’t recover from this problem (build.log.3).  

[build2.log](https://github.com/iraf-community/iraf/files/10900480/build2.log)
[build3.log](https://github.com/iraf-community/iraf/files/10900481/build3.log)
![Screenshot 2023-03-06 at 10 40 44](https://user-images.githubusercontent.com/127039683/223173098-e1b57202-9ccb-4842-a01b-cb2175c72fcd.jpg)

**Reply by olebole** [reply]
There is still the `xc: Command not found` line in the build log, which means that `xc` is not in your path. You could try to add `/Users/irs/.irf/bin` to you PATH before re-running `make sysgen 2>&1 | tee build.log`, so that it can be found. Check that it can be found with `which xc`.

**Reply by ian-smail** [reply]
yes - i've confirmed that the PATH variable contains /Users/irs/.iraf/bin and a "which xc" command shows that it is being found.   but then running the make sysgen in that shell results in the error.   

so - here is the error from running the make and then the PATH and "which/file xc" showing that the shell can see xc and it appears to be an arm64 executable.   so at least the shell from which the make is being run can see xc...

![Screenshot 2023-03-06 at 17 09 16](https://user-images.githubusercontent.com/127039683/223181284-1ecd1109-f53c-4354-9506-a0f517695de9.jpg)

**Reply by olebole** [reply]
Hmm. Can you execute **xc** from the command line? 
```
$ xc -V
IRAFNET XC V2.4 Jan 21 2010
```

**Reply by olebole** [reply]
Can you also check that the `$iraf` and `$IRAFARCH` shell variables are set properly?

**Reply by ian-smail** [reply]
bash-3.2$ xc -V
IRAFNET XC V2.4 Jan 21 2010
clang: error: no such file or directory: 'libmain.o'
clang: error: no such file or directory: 'libex.a'
clang: error: no such file or directory: 'liblapack.a'
clang: error: no such file or directory: 'libvops.a'
clang: error: no such file or directory: 'libfftpack.a'
clang: error: no such file or directory: 'libVOTable.a'
clang: error: no such file or directory: 'libcfitsio.a'

echo $IRAFARCH
macos64

 echo $iraf
/usr/local/iraf/iraf-2.17/

...that last one is wrong - given i'm now doing the build on /Users/irs/Desktop/iraf-2.17

but resetting the variable to the correct directory and rerunning doesn't fix the make error...

HOWEVER - i think the sensible thing is that i should remove all the various partial builds, reboot my machine and start this from scratch in as clean a way as possible - and then get back to you...  you've spent enough time on this already.

thanks - ian

**Comment by ian-smail** [comment]
mmmhh.... that was less successful than i'd hoped.  it's still barfing in the same place - and still complaining about xc not existing - even though it clearly does in the relevant path.

here are screen grabs before the make sysgen, after the make fails and also in case it is relevant the contents of the lib/hlib directories - from the make file it appears that there should be many more files copied into here that don't appear (stuff from gdev and sgidev)?  

i should note that i'm not a native bash shell user (i'm still on tcsh) - but for these experiments i've changed the default shell and deleted all the .bash* files before launching...  

![before_make](https://user-images.githubusercontent.com/127039683/223491681-78969baf-7142-411e-9b88-9e44e07f02a1.jpg)
![after_make](https://user-images.githubusercontent.com/127039683/223491713-df077c99-bde8-44b2-9963-8117ebda7c3a.jpg)
[build.log](https://github.com/iraf-community/iraf/files/10911886/build.log)
![hlib_and_lib](https://user-images.githubusercontent.com/127039683/223491745-b78c615f-9498-4b43-af2a-6c5af67b4bb3.jpg)

**Reply by olebole** [reply]
> i should note that i'm not a native bash shell user (i'm still on tcsh) - but for these experiments i've changed the default shell and deleted all the .bash* files before launching...

That may be the cause... in IRAF 2.17, everything switched to Bash, and so removing all Bash dotfiles may have some unexpected consequences.

**Reply by ian-smail** [reply]
i'm running with the default bash account - so it has a minimal path.   i've tried building before the IRAF added stuff to the .bash files and afterwards and neither seems to fix this...   so i guess i'm stuck.   thanks for your help - ian

**Reply by olebole** [reply]
I just made some changes in the code to make the build process more robust. Maybe you could try out this test version:

https://github.com/olebole/iraf/archive/refs/heads/mkpkg-vars.tar.gz

Building on a pristine Mac succeded in a test. If it fails on your computer, please attach the build log.
For discussion, this is Pull Request #262.

**Comment by ian-smail** [comment]
dear oli - 

apologies for the silence - i was fixing something else... 

anyway - that works!   10 minutes to build it - but i've got a working cl prompt and can display dev$pix on ds9.    the test run reported 2 skipped and 1 failed - but i couldn't work out what the failure was - and its obviously not mission-critical.

thank you very much and sorry for all the work - but this will make my life a lot easier!

ian
